terraso_backend/apps/project_management/permission_rules.py

```diff
@@ -12,137 +12,118 @@
 #
 # You should have received a copy of the GNU Affero General Public License
 # along with this program. If not, see https://www.gnu.org/licenses/.
-import rules
-
-from apps.project_management.collaboration_roles import ProjectRole
-
-
-@rules.predicate
-def allowed_to_change_project(user, project):
-    return project.is_manager(user)
-
-
-@rules.predicate
-def allowed_to_add_site_to_project(user, project):
-    return project.is_manager(user) or project.is_contributor(user)
+from dataclasses import dataclass
 
+import rules
 
-@rules.predicate
-def allowed_to_update_site(user, site):
-    if site.owned_by_user:
-        return site.owner == user
-    return site.project.is_manager(user) or site.project.is_contributor(user)
+from apps.project_management.models import Project, Site, SiteNote
 
 
-@rules.predicate
-def allowed_to_delete_site(user, site):
-    if site.owned_by_user:
-        return site.owner == user
-    return site.project.is_manager(user)
+@dataclass
+class Context:
+    # these are the target entity(s) the operation is occurring on. they will always be
+    # referentually consistent (i.e., if the context has a site this will always be the site's
+    # project)
+    project: Project = None
+    site: Site = None
+    site_note: SiteNote = None
 
+    # these are the source entities which may also be part of the operation (i.e., the site which
+    # is being added to a new owner, and that site's current project if it has one)
+    source_project: Project = None
+    source_site: Site = None
 
-@rules.predicate
-def allowed_to_update_site_settings(user, site):
-    if site.owned_by_user:
-        return site.owner == user
-    return site.project.is_manager(user)
-
-
-@rules.predicate
-def allowed_to_delete_project(user, project):
-    return project.is_manager(user)
+    def __post_init__(self):
+        if self.site_note:
+            self.site = self.site_note.site
+        if self.site:
+            self.project = self.site.project
+        if self.source_site:
+            self.source_project = self.source_site.project
 
 
 @rules.predicate
-def allowed_to_add_to_project(user, project):
-    return project.is_manager(user)
+def allowed_to_create(user, context):
+    # (all logged-in users are allowed base creation actions)
+    return True
 
 
 @rules.predicate
-def allowed_to_archive_project(user, project):
-    return project.is_manager(user)
+def allowed_to_manage_project(user, context):
+    return context.project.is_manager(user)
 
 
 @rules.predicate
-def allowed_to_add_member_to_project(user, context):
-    project = context["project"]
-    requester_membership = context["requester_membership"]
+def allowed_to_be_project_member(user, context):
```

Comment:
opinion: i find `allowed_to_be_project_member` confusing as a permission name? the implementation check reads to me like it is checking whether the user is currently a member of the project, while the name makes me think it is checking for _any_ user whether they _could_ be a member of the project

Reply:
I agree that this isn't a great name -- am open to suggestions.